### PR TITLE
Detect undefined variables or filters

### PR DESCRIPTION
**Pull request by apeckham**
This adds `strict_variables` and `strict_filters` options to render. When true, `render` collects a list of undefined variables/filters in #errors, and `render!` raises when an undefined variable/filter is used.

Example:

```ruby
template = Liquid::Template.parse("{{x}} {{y}} {{z.a}} {{z.b}}")
template.render({ 'x' => 1, 'z' => { 'a' => 2 } }, { strict_variables: true })
template.errors
#=> [#<Liquid::MissingVariable: Liquid error: missing variable y>, #<Liquid::MissingVariable: Liquid error: missing variable b in {"a"=>2}>]
```

This lets us write unit tests for our templates. Related:
- #685 Get all variables used in a template
- #490 Fail on nonexistent variable dereferencing
- #422 Invalid filters are ignored without warning

Thanks @Jeiwan for working on this.

The loop over [global_filter, exception_handler, strict_variables, strict_filters] in template.rb is to fix a Rubocop cyclomatic complexity error.

Tests included. Will you consider merging? Happy to make changes.

### CODE REVIEW

File: lib/liquid/template.rb [hunk 1]
```diff
@@ -183,10 +183,9 @@ def render(*args)
 
         context.add_filters(options[:filters]) if options[:filters]
 
-        context.global_filter = options[:global_filter] if options[:global_filter]
-
-        context.exception_handler = options[:exception_handler] if options[:exception_handler]
-
+        %i(global_filter exception_handler strict_variables strict_filters).each do |option|
+          context.send(:"#{option}=", options[option]) if options[option]
+        end
```
Comment (pushrax) [inline lib/liquid/template.rb]: This is taking DRY too far IMO.

**Comment by pushrax** [comment]
Great work, I think this will be useful for many people. I'll give this PR an actual review tonight before passing judgement.

**Comment by fw42** [comment]
@hkdsun, can you review this as well please?

**Comment by apeckham** [comment]
When strict_variables is true `{% include "foo" %}` raises "missing variable foo" on context.rb:211. I wrote a failing test and we'll investigate a fix.

**Comment by apeckham** [comment]
Actually, I'm not sure what the correct behavior should be in the above situation, after reading include_tag_test. Let us know your thoughts!

File: lib/liquid/context.rb [hunk 2]
```diff
@@ -27,7 +27,7 @@ def initialize(environments = {}, outer_scope = {}, registers = {}, rethrow_erro
       @this_stack_used = false
 
       if rethrow_errors
-        self.exception_handler = ->(e) { raise }
+        self.exception_handler = ->(e) { raise e }
```
Comment (fw42) [inline lib/liquid/context.rb]: Why this change? I don't think this is equivalent.

File: lib/liquid/context.rb [hunk 3]
```diff
@@ -207,6 +207,8 @@ def find_variable(key)
     def lookup_and_evaluate(obj, key)
       if (value = obj[key]).is_a?(Proc) && obj.respond_to?(:[]=)
         obj[key] = (value.arity == 0) ? value.call : value.call(self)
+      elsif strict_variables && obj.respond_to?(:has_key?) && !obj.has_key?(key)
+        raise Liquid::MissingVariable, "missing variable #{key}"
```
Comment (fw42) [inline lib/liquid/context.rb]: Instead of "missing", can we say "undefined"?

File: lib/liquid/template.rb (hunk 1 above)
Comment (fw42) [inline lib/liquid/template.rb]: Maybe make a new method somewhere? `parsee_options` or `configure_context` or whatever?

File: lib/liquid/variable_lookup.rb [hunk 4]
```diff
@@ -54,6 +54,8 @@ def evaluate(context)
         elsif @command_flags & (1 << i) != 0 && object.respond_to?(key)
           object = object.send(key).to_liquid
 
+        elsif context.strict_variables
+          raise Liquid::MissingVariable, "missing variable #{key} in #{object}"
```
Comment (fw42) [inline lib/liquid/variable_lookup.rb]: I'm worried about the `in #{object}` part here. That looks like it could potentially leak sensitive information into the rendered output. Can we remove that?

File: test/integration/strainer_test.rb [hunk 5]
```diff
@@ -0,0 +1,15 @@
+require 'test_helper'
+
+class StrainerTest < Minitest::Test
+  include Liquid
+
+  def test_add_filter_when_wrong_filter_class
```
Comment (fw42) [inline test/integration/strainer_test.rb]: This test looks like a unit test to me, not an integration test... Why is it in `test/integration/`?

File: README.md [hunk 6]
```diff
@@ -73,3 +73,30 @@ This is useful for doing things like enabling strict mode only in the theme edit
 
 It is recommended that you enable `:strict` or `:warn` mode on new apps to stop invalid templates from being created.
 It is also recommended that you use it in the template editors of existing apps to give editors better error messages.
+
+### Undefined variables and filters
+
+By default, the renderer doesn't raise or in any other way notify you if some variables or filters are missing, i.e. not passed to the `render` method. You can improve this situation by passing `strict_variables: true` and/or `strict_filters: true` options to `render` method. When one of these options is set to true, all missing errors will be stored in `errors` array of a `Liquid::Template` instance.
```
Comment (fw42) [inline README.md]: can you add some line breaks here?

File: lib/liquid/block_body.rb [hunk 7]
```diff
@@ -76,6 +76,9 @@ def render(context)
           end
         rescue MemoryError => e
           raise e
+        rescue MissingVariable, MissingDropMethod, MissingFilter => e
+          context.handle_error(e, token.line_number)
+          output << nil
```
Comment (fw42) [inline lib/liquid/block_body.rb]: Why is this necessary? Do any tests fail if you remove this line?

File: lib/liquid/strainer.rb [hunk 8]
```diff
@@ -26,7 +26,7 @@ class << self
     end
 
     def self.add_filter(filter)
-      raise ArgumentError, "Expected module but got: #{f.class}" unless filter.is_a?(Module)
+      raise ArgumentError, "Expected module but got: #{filter.class}" unless filter.is_a?(Module)
```
Comment (fw42) [inline lib/liquid/strainer.rb]: This was a bug that you are fixing? Can you add a regression test?

**Comment by fw42** [comment]
This looks pretty cool! I'm tempted to even make this the default behaviour in strict mode.

File: lib/liquid/template.rb (hunk 1 above)
Comment (Jeiwan) [inline lib/liquid/template.rb]: Maybe `apply_options_to_context`?

File: lib/liquid/context.rb (hunk 2 above)
Comment (Jeiwan) [inline lib/liquid/context.rb]: We want to rethrow the same exception, `raise` without parameters throws `RuntimeError` without a message.

File: lib/liquid/variable_lookup.rb (hunk 4 above)
Comment (Jeiwan) [inline lib/liquid/variable_lookup.rb]: The idea is not to print exception messages into the rendered output. This message is only available in `errors` method.
And there's also a problem, that we cannot get full variable path here, only variable name and current scope. So I decided to add the scope to the message for developers to know where exactly the missing variable is.

File: test/integration/strainer_test.rb (hunk 5 above)
Comment (Jeiwan) [inline test/integration/strainer_test.rb]: Hmmm... yep, will move it to unit tests.

File: lib/liquid/block_body.rb (hunk 7 above)
Comment (Jeiwan) [inline lib/liquid/block_body.rb]: We didn't want to print exception messages to the output. Actually this approach was implemented in https://github.com/Shopify/liquid/issues/490 so I also decided to use it.

File: lib/liquid/strainer.rb (hunk 8 above)
Comment (Jeiwan) [inline lib/liquid/strainer.rb]: Yep, it's a bug I happened to come across. Decided to fix it along the way. The test is already there, `test/integration/strainer_test.rb` I'll just move it to unit tests, as discussed above.

File: lib/liquid/context.rb (hunk 3 above)
Comment (Jeiwan) [inline lib/liquid/context.rb]: Fixed

File: README.md (hunk 6 above)
Comment (Jeiwan) [inline README.md]: Done

**Comment by hkdsun** [comment]
This looks good to me after the latest change :+1:

I can also see it being the default behaviour in strict mode, since we encourage strict mode as a way of validating the correctness of templates

File: lib/liquid/context.rb (hunk 2 above)
Comment (fw42) [inline lib/liquid/context.rb]: Do any tests fail if you revert this change here?

Comment (nickpearson) [inline lib/liquid/context.rb]: I don't believe adding the `e` changes the behavior. Calling `raise` inside a `rescue` block re-raises the rescued error. For example:

```ruby
begin; raise ArgumentError, 'test'; rescue; raise; end
# and
begin; raise ArgumentError, 'test'; rescue => e; raise e; end
# both result in => ArgumentError: test
```

Comment (fw42) [inline lib/liquid/context.rb]: yeah that's what I thought too. Not sure about the semantics if it's with a lambda, but `raise e` and `raise` is usually not equivalent (one re-raises, the other raises a new exception (with a new backtrace)).

Comment (nickpearson) [inline lib/liquid/context.rb]: It seems `raise` pickups up the error even when called in a lambda, as long as the lambda is called from within the `rescue` block:

```ruby
h1 = ->(e){ raise }
h2 = ->(e){ raise e }
begin; raise ArgumentError, 'test'; rescue => e; h1.(e); end
begin; raise ArgumentError, 'test'; rescue => e; h2.(e); end
# both still result in => ArgumentError: test
```

File: README.md [hunk 9]
```diff
@@ -73,3 +73,32 @@ This is useful for doing things like enabling strict mode only in the theme edit
 
 It is recommended that you enable `:strict` or `:warn` mode on new apps to stop invalid templates from being created.
 It is also recommended that you use it in the template editors of existing apps to give editors better error messages.
+
+### Undefined variables and filters
+
+By default, the renderer doesn't raise or in any other way notify you if some variables or filters are missing, i.e. not passed to the `render` method.
+You can improve this situation by passing `strict_variables: true` and/or `strict_filters: true` options to `render` method.
```
Comment (fw42) [inline README.md]: "to **the** render method"

File: README.md [hunk 10]
```diff
@@ -73,3 +73,32 @@ This is useful for doing things like enabling strict mode only in the theme edit
...
+When one of these options is set to true, all missing errors will be stored in `errors` array of a `Liquid::Template` instance.
```
Comment (fw42) [inline README.md]: maybe "all errors about undefined variables and undefined filters" instead of "all missing errors"?

File: README.md [hunk 11]
```diff
@@ -73,3 +73,32 @@ This is useful for doing things like enabling strict mode only in the theme edit
...
+Here are some examples:
+
+```ruby
+template = Liquid::Template.parse("{{x}} {{y}} {{z.a}} {{z.b}}")
+template.render({ 'x' => 1, 'z' => { 'a' => 2 } }, { strict_variables: true })
+template.errors
+#=> [#<Liquid::MissingVariable: Liquid error: missing variable y>, #<Liquid::MissingVariable: Liquid error: missing variable b in {"a"=>2}>]
```
Comment (fw42) [inline README.md]: Example still has the `MissingVariable`. I think you renamed that. Can you update here?

File: lib/liquid/context.rb [hunk 12]
```diff
@@ -207,6 +207,8 @@ def find_variable(key)
     def lookup_and_evaluate(obj, key)
       if (value = obj[key]).is_a?(Proc) && obj.respond_to?(:[]=)
         obj[key] = (value.arity == 0) ? value.call : value.call(self)
+      elsif strict_variables && obj.respond_to?(:has_key?) && !obj.has_key?(key)
```
Comment (fw42) [inline lib/liquid/context.rb]: I think this should be `key?`, not `has_key?` (see https://github.com/Shopify/liquid/pull/593)

File: lib/liquid/variable_lookup.rb [hunk 13]
```diff
@@ -54,6 +54,8 @@ def evaluate(context)
         elsif @command_flags & (1 << i) != 0 && object.respond_to?(key)
           object = object.send(key).to_liquid
 
+        elsif context.strict_variables
```
Comment (fw42) [inline lib/liquid/variable_lookup.rb]: Hm how about moving this into the `else`? Like
```ruby
else
  return nil unless context.strict_variables
  raise Liquid::UndefinedVariable, "undefined variable #{key} in #{object}"
end
```
I think that would be a bit clearer?

File: lib/liquid/variable_lookup.rb [hunk 14]
```diff
@@ -54,6 +54,8 @@ def evaluate(context)
...
+          raise Liquid::UndefinedVariable, "undefined variable #{key} in #{object}"
```
Comment (fw42) [inline lib/liquid/variable_lookup.rb]: I'm still worried about including `#{object}` in the string. I don't see the benefit, but I do see potential leaks.

File: lib/liquid/context.rb (hunk 12 above)
Comment (hkdsun) [inline lib/liquid/context.rb]: Sorry. I deleted a comment from this line by accident. 

@fw42 had said:
>I think this should be key?, not has_key? (see #593)

File: lib/liquid/context.rb (hunk 2 above)
Comment (Jeiwan) [inline lib/liquid/context.rb]: Yep, it fails without `e`:
```
  1) Failure:
TemplateTest#test_undefined_filters_raise [/Users/Jeiwan/Work/liquid/liquid/test/integration/template_test.rb:296]:
[Liquid::UndefinedFilter] exception expected, not
Class: <RuntimeError>
Message: <"">
---Backtrace---
/Users/Jeiwan/Work/liquid/liquid/lib/liquid/context.rb:30:in `block in initialize'
/Users/Jeiwan/Work/liquid/liquid/lib/liquid/context.rb:84:in `handle_error'
/Users/Jeiwan/Work/liquid/liquid/lib/liquid/strainer.rb:52:in `invoke'
/Users/Jeiwan/Work/liquid/liquid/lib/liquid/context.rb:103:in `invoke'
/Users/Jeiwan/Work/liquid/liquid/lib/liquid/variable.rb:78:in `block in render'
/Users/Jeiwan/Work/liquid/liquid/lib/liquid/variable.rb:76:in `each'
/Users/Jeiwan/Work/liquid/liquid/lib/liquid/variable.rb:76:in `inject'
/Users/Jeiwan/Work/liquid/liquid/lib/liquid/variable.rb:76:in `render'
/Users/Jeiwan/Work/liquid/liquid/lib/liquid/block_body.rb:93:in `render_node'
/Users/Jeiwan/Work/liquid/liquid/lib/liquid/profiler/hooks.rb:5:in `block in render_node_with_profiling'
/Users/Jeiwan/Work/liquid/liquid/lib/liquid/profiler.rb:80:in `profile_node_render'
/Users/Jeiwan/Work/liquid/liquid/lib/liquid/profiler/hooks.rb:4:in `render_node_with_profiling'
/Users/Jeiwan/Work/liquid/liquid/lib/liquid/block_body.rb:72:in `block in render'
/Users/Jeiwan/Work/liquid/liquid/lib/liquid/block_body.rb:59:in `each'
/Users/Jeiwan/Work/liquid/liquid/lib/liquid/block_body.rb:59:in `render'
/Users/Jeiwan/Work/liquid/liquid/lib/liquid/template.rb:196:in `block in render'
/Users/Jeiwan/Work/liquid/liquid/lib/liquid/template.rb:230:in `with_profiling'
/Users/Jeiwan/Work/liquid/liquid/lib/liquid/template.rb:195:in `render'
/Users/Jeiwan/Work/liquid/liquid/lib/liquid/template.rb:208:in `render!'
/Users/Jeiwan/Work/liquid/liquid/test/integration/template_test.rb:297:in `block in test_undefined_filters_raise'
---------------

519 runs, 1277 assertions, 1 failures, 0 errors, 0 skips
rake aborted!
```

This happens only with `UndefinedFilter` exception because we don't actually raise it, but just handle it with `@context.handle_error` here: https://github.com/Shopify/liquid/pull/691/files#diff-cc7b2b233f09fa41b38a4a44e7c5bb3cR52 So this `exception_handler` executes not inside a `rescue`.
And we can't raise it because we don't want to interrupt the filter chain. So if we have filters like these: `{'foo' | undefinedfilter1 | upcase | undefinedfilter2}` we just want to end up with `FOO` and skip all undefined filters.

File: README.md (hunk 9 above)
Comment (Jeiwan) [inline README.md]: Fixed

File: README.md (hunk 10 above)
Comment (Jeiwan) [inline README.md]: Fixed

File: README.md (hunk 11 above)
Comment (Jeiwan) [inline README.md]: Sorry, missed this. Fixed now.

File: lib/liquid/variable_lookup.rb (hunk 13 above)
Comment (Jeiwan) [inline lib/liquid/variable_lookup.rb]: :+1:

File: lib/liquid/variable_lookup.rb (hunk 14 above)
Comment (Jeiwan) [inline lib/liquid/variable_lookup.rb]: Here's an example where it can be useful:
```ruby
t = Template.parse("{{a}} {{b}} {{x.a}} {{x.b}}")
t.render({'a' => 1, 'x' => { 'a' => 2}}, { strict_variables: true})
# => "1  2 "
 => [#<Liquid::UndefinedVariable: Liquid error: undefined variable b>, #<Liquid::UndefinedVariable: Liquid error: undefined variable b in {"a"=>2}>]
```
The second error message gives more information to a developer where the missing variable is. If we remove `in #{object}` part we will end up with two identical errors messages:
```ruby
# => [#<Liquid::UndefinedVariable: Liquid error: undefined variable b>, #<Liquid::UndefinedVariable: Liquid error: undefined variable b>]
```

File: lib/liquid/context.rb (hunk 12 above)
Comment (Jeiwan) [inline lib/liquid/context.rb]: Fixed

File: lib/liquid/context.rb (hunk 2 above)
Comment (fw42) [inline lib/liquid/context.rb]: Hm I actually think it would be more intuitive to interrupt the filter chain. Why do you think it's better not to?

Comment (pushrax) [inline lib/liquid/context.rb]: Imagine `x | undefined | escape`. However I think I agree that the filter chain should be interrupted **and** the result forced to `nil`.

Comment (Jeiwan) [inline lib/liquid/context.rb]: I think it's better to not interrupt the filter chain. I allows us to get the most successful render result, i.e. all defined filters will be applied and undefined filters will be just skipped. At least this is what we have now, without the changes in this PR.
But it's up to you to decide, which approach is better here.

Comment (pushrax) [inline lib/liquid/context.rb]: I would argue that with `strict_filters` enabled you should not be able to reference an undefined filter. Unlike variables (which might not be defined yet since they are controlled by the user), filters are static and if undefined must be a typo.

**Comment by Jeiwan** [comment]
@fw42 @nickpearson @pushrax Check this commit out: https://github.com/urbandictionary/liquid/commit/f42518c86a40828000f712dc6c836cd327ebf99d
I reverted the change in `exception_handler`, and now interrupt the filter chain and return `nil` to the output when `UndefinedFilter` exception is risen. There's one issue though: `errors` now contains only the first undefined filter in the chain, not all of them.

File: lib/liquid/variable_lookup.rb [hunk 15]
```diff
@@ -55,9 +55,11 @@ def evaluate(context)
           object = object.send(key).to_liquid
 
           # No key was present with the desired value and it wasn't one of the directly supported
-          # keywords either. The only thing we got left is to return nil
+          # keywords either. The only thing we got left is to return nil or
+          # raise an exception if `strict_variables` option is set to true
         else
-          return nil
+          return nil unless context.strict_variables
+          raise Liquid::UndefinedVariable, "undefined variable #{key} in #{object}"
```
Comment (fw42) [inline lib/liquid/variable_lookup.rb]: I see your point here about how this could be useful for developers, but I still disagree and at least in Shopify's use-case (where we don't necessarily trust the author of the template), I think exposing `object` is too dangerous.

**Comment by fw42** [comment]
I like this. If you remove the `object` in the error message, I'm happy to merge this.

@pushrax, :+1:?

**Comment by nickpearson** [comment]
Only capturing one error seems fine to me, as that's how any other programming language (that I'm aware of) works -- it quits on the first unhandled error since the application state is indeterminate at that point and cannot safely continue.

File: lib/liquid/template.rb [hunk 16]
```diff
@@ -181,12 +181,7 @@ def render(*args)
 
         registers.merge!(options[:registers]) if options[:registers].is_a?(Hash)
 
-        context.add_filters(options[:filters]) if options[:filters]
-
-        context.global_filter = options[:global_filter] if options[:global_filter]
-
-        context.exception_handler = options[:exception_handler] if options[:exception_handler]
-
+        apply_options_to_context(context, options)
```
Comment (pushrax) [inline lib/liquid/template.rb]: Good refactor.

File: lib/liquid/variable_lookup.rb (hunk 15 above)
Comment (pushrax) [inline lib/liquid/variable_lookup.rb]: Agree completely.

**Comment by Jeiwan** [comment]
I removed `in #{object}` from the message and improved README (added render results with comments to the examples).

**Comment by pushrax** [comment]
Thanks for this contribution!

**Comment by Jeiwan** [comment]
Great! Thank you for merging it.

**Comment by gauravmc** [comment]
FYI, this caused a bunch of `NoMethodError·undefined method `strict_variables' for nil:NilClass` when latest ref of the gem was tried in Shopify core.

File: lib/liquid/drop.rb [hunk 17]
```diff
@@ -24,8 +24,9 @@ class Drop
     attr_writer :context
 
     # Catch all for the method
-    def liquid_method_missing(_method)
-      nil
+    def liquid_method_missing(method)
+      return nil unless @context.strict_variables
```
Comment (dylanahsmith) [inline lib/liquid/drop.rb]: @gauravmc (regarding https://github.com/Shopify/liquid/pull/691#issuecomment-188461364) more specifically a drop doesn't always have the `@context` set.  This is the line that was causing the NoMethodError exceptions.
